modules/ip/cli/icmp.c

```diff
@@ -0,0 +1,220 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2024 Christophe Fontaine
+
+#include <gr_api.h>
+#include <gr_cli.h>
+#include <gr_ip4.h>
+#include <gr_net_types.h>
+
+#include <ecoli.h>
+
+#include <errno.h>
+#include <signal.h>
+#include <sys/queue.h>
+#include <unistd.h>
+
+struct ipv4_hdr {
+	union {
+		uint8_t version_ihl; /**< version and header length */
+		struct {
+#if RTE_BYTE_ORDER == RTE_LITTLE_ENDIAN
+			uint8_t ihl : 4; /**< header length */
+			uint8_t version : 4; /**< version */
+#elif RTE_BYTE_ORDER == RTE_BIG_ENDIAN
+			uint8_t version : 4; /**< version */
+			uint8_t ihl : 4; /**< header length */
+#endif
+		};
+	};
+	uint8_t type_of_service; /**< type of service */
+	uint16_t total_length; /**< length of packet */
+	uint16_t packet_id; /**< packet ID */
+	uint16_t fragment_offset; /**< fragmentation offset */
+	uint8_t time_to_live; /**< time to live */
+	uint8_t next_proto_id; /**< protocol ID */
+	uint16_t hdr_checksum; /**< header checksum */
+	uint32_t src_addr; /**< source address */
+	uint32_t dst_addr; /**< destination address */
+} __attribute__((packed));
+
+static const char *ICMP_DEST_UNREACHABLE[] = {
+	"net unreachable",
+	"host unreachable",
+	"protocol unreachable"
+	"port unreachable",
+	"fragmentation needed and DF set",
+	"source route failed",
+};
+
+static bool stop = false;
+
+static void sighandler(int) {
+	stop = true;
+}
+
+static cmd_status_t icmp_request(
+	const struct gr_api_client *c,
+	struct gr_ip4_icmp_request *req,
+	uint16_t msdelay,
+	uint16_t count,
+	bool mode_traceroute
+) {
+	cmd_status_t ret_code = CMD_SUCCESS;
+	struct gr_ip4_icmp_response *resp;
+	void *resp_ptr = NULL;
+	int timeout;
+	int ret;
+
+	for (int i = mode_traceroute; i < count && stop == false; i++) {
+		req->ttl = (mode_traceroute ? i : 64);
+		req->sequence_number = i;
+
+		if (gr_api_client_send_recv(c, GR_IP4_ICMP_REQUEST, sizeof(*req), req, NULL) < 0) {
+			ret_code = CMD_ERROR;
+			goto exit;
+		}
+
+		timeout = 50;
+		resp_ptr = NULL;
+		do {
+			usleep(10000);
```

Comment:
We'll never get the icmp response as we just sent the request, so it was a good tradeoff to wait initially, instead of waiting at the end of the loop.

Reply:
But shouldn't you usleep() for the duration of the msdelay?

Reply:
we have a fixed timeout of 500ms.
msdelay is the time between icmp probes.
Do you prefer to also have a configurable timeout ?